flink-runtime/src/main/java/org/apache/flink/runtime/util/EnvironmentInformation.java

```diff
@@ -41,42 +45,143 @@
+		private Instant gitCommitTime = DEFAULT_TIME_INSTANT;
+		private String  gitCommitTimeStr = DEFAULT_TIME_STRING;
+
+		private static final String PROP_FILE = ".flink-runtime.version.properties";
+
+		private static final String FAIL_MESSAGE = "The file " + PROP_FILE + " has not been generated correctly. You MUST run 'mvn generate-sources' in the flink-runtime module.";
+
+		private String getProperty(Properties properties, String key, String defaultValue) {
+			String value = properties.getProperty(key);
+			if (value == null || value.charAt(0) == '$') {
+				return defaultValue;
 			}
-		} catch (Throwable t) {
-			if (LOG.isDebugEnabled()) {
-				LOG.debug("Cannot determine code revision: Unable to read version property file.", t);
-			} else {
-				LOG.info("Cannot determine code revision: Unable to read version property file.");
+			return value;
+		}
+
+		public Versions() {
+			ClassLoader classLoader = EnvironmentInformation.class.getClassLoader();
+			try (InputStream propFile = classLoader.getResourceAsStream(PROP_FILE)) {
+				if (propFile != null) {
+					Properties properties = new Properties();
+					properties.load(propFile);
+
+					projectVersion = getProperty(properties, "project.version", UNKNOWN);
+					scalaVersion = getProperty(properties, "scala.binary.version", UNKNOWN);
+
+					gitCommitId = getProperty(properties, "git.commit.id", UNKNOWN_COMMIT_ID);
+					gitCommitIdAbbrev = getProperty(properties, "git.commit.id.abbrev", UNKNOWN_COMMIT_ID_ABBREV);
+
+					// This is to reliably parse the datetime format configured in the git-commit-id-plugin
+					DateTimeFormatter gitDateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ssZ");
+
+					// Default format is in Berlin timezone because that is where Flink originated.
+					DateTimeFormatter BERLIN_DATE_TIME = DateTimeFormatter.ISO_OFFSET_DATE_TIME.withZone(ZoneId.of("Europe/Berlin"));
+
+					try {
+						String propGitCommitTime = getProperty(properties, "git.commit.time", DEFAULT_TIME_STRING);
+						gitCommitTime = gitDateTimeFormatter.parse(propGitCommitTime, Instant::from);
+						gitCommitTimeStr = BERLIN_DATE_TIME.format(gitCommitTime);
+
+						String propGitBuildTime  = getProperty(properties, "git.build.time", DEFAULT_TIME_STRING);
+						gitBuildTime = gitDateTimeFormatter.parse(propGitBuildTime, Instant::from);
+						gitBuildTimeStr = BERLIN_DATE_TIME.format(gitBuildTime);
+					} catch (DateTimeParseException dtpe) {
+						LOG.error("{} : {}", FAIL_MESSAGE, dtpe);
+						throw new IllegalStateException(FAIL_MESSAGE);
+					}
+				}
+			}
+			catch(IOException ioe) {
+				LOG.info("Cannot determine code revision: Unable to read version property file.: {}", ioe.getMessage());
 			}
 		}
-		
-		return new RevisionInformation(revision, commitDate);
+	}
+
+	private static final class VersionsHolder {
+		private static final Versions INSTANCE = new Versions();
+	}
+
+	public static Versions getVersionsInstance() {
```

Comment:
private?

Reply:
(since `Versions` is private anyway so you couldn't do anything interesting with it)

Reply:
Ah, yes. I'll fix that.